crates/red_knot/build.rs

```diff
@@ -0,0 +1,90 @@
+//! Build script to package our vendored typeshed files
+//! into a zip archive that can be included in the Ruff binary.
+//!
+//! This script should be automatically run at build time
+//! whenever the script itself changes, or whenever any files
+//! in `crates/red_knot/vendor/typeshed` change.
+
+use std::error::Error;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::Path;
+
+use walkdir::{DirEntry, WalkDir};
+use zip::result::ZipResult;
+use zip::write::{FileOptions, ZipWriter};
+use zip::CompressionMethod;
+
+const TYPESHED_SOURCE_DIR: &str = "vendor/typeshed";
+
+// NB: This is .gitignored; make sure to change the .gitignore entry
+// if you change this location
+const TYPESHED_ZIP_LOCATION: &str = "vendor/zipped_typeshed.zip";
+
+/// Recursively zip the contents of an entire directory.
+///
+/// This routine is adapted from a recipe at
+/// <https://github.com/zip-rs/zip-old/blob/5d0f198124946b7be4e5969719a7f29f363118cd/examples/write_dir.rs>
+fn zip_dir(
+    directory_iterator: &mut impl Iterator<Item = DirEntry>,
+    prefix: &str,
+    writer: File,
+) -> ZipResult<()> {
+    let mut zip = ZipWriter::new(writer);
+
+    let options = FileOptions::default()
+        .compression_method(CompressionMethod::Zstd)
+        .unix_permissions(0o644);
+
+    let mut buffer = Vec::new();
+    for entry in directory_iterator {
+        let path = entry.path();
+        let name = path
```

Comment:
Maybe rename `name` to `relative_path`. Because I think it isn't just the file's name, it's actually the file paths, just relative to the `root` directory.